magda-web-client/src/helpers/ChartDatasetEncoder.js

```diff
@@ -155,173 +395,30 @@ class ChartDatasetEncoder {
-                ...field,
-                idx: indexOf(headerRow, key),
-                name: key,
-                label: startCase(key),
-                category: !field.time && !field.numeric,
-                isAggr: false
-            })
+        this.fields = preProcessFields(
+            Object.keys(this.data[0]),
+            this.distribution
         );
-        //--- filter out fields that cannot be located in CSV data. VisualInfo outdated maybe?
-        newFields = filter(newFields, item => item.idx !== -1);
-        if (!newFields.length) {
-            //--- we will not exit but make our own guess
-            newFields = map(headerRow, (headerName, idx) => {
-                return this.predictInitDataFieldType(headerName, {
-                    idx,
-                    name: headerName,
-                    label: startCase(headerName),
-                    category: true,
-                    time: false,
-                    numeric: false,
-                    isAggr: false
-                });
-            });
-        }
-        newFields = filter(newFields, item => trim(item.name) !== "");
-        if (!newFields.length) {
-            throw new Error("The data file contains no non-empty header.");
-        }
-        this.fields = newFields;
-        return newFields;
-    }
 
-    preProcessData() {
-        if (!this.data || this.data.length < 1)
-            throw new Error("The data file loaded is empty.");
-        this.preProcessFields(Object.keys(this.data[0]));
-        //--- if only one non-numeric column, add new column by count
-        if (this.fields.length === 1 && !this.fields[0].numeric) {
-            const newFieldName = "count_" + Math.random();
-            this.fields.push({
-                idx: 1,
-                name: newFieldName,
-                label: "Count",
-                time: false,
-                numeric: true,
-                category: false,
-                isAggr: true,
-                isAggrDone: true
-            });
-            this.data = this.groupBy(
-                this.data,
-                newFieldName,
-                aggregators.count,
-                [this.fields[0].name]
-            );
-        }
-        //--- if unfortunately no numeric cols, present data by selected col's count
-        if (!this.getNumericColumns().length) {
-            const newFieldName = "count_" + Math.random();
-            this.fields.push({
-                idx: 1,
-                name: newFieldName,
-                label: "Count",
-                time: false,
-                numeric: true,
-                category: false,
-                isAggr: true,
-                isAggrDone: false
-            });
-            //--- we cann't generate coutn data here yet as we don't know user's selection
-        }
+        //--- Present data by selected col's count
+
+        const newFieldName = "count";
+        this.fields.push({
+            idx: 1,
```

Comment:
Since we are adding `count` for all chart, we'd better make sure it's unique among all columns (it (and probably `name`) will be used as column reference by eCharts).
We can set to: 
```
idx: this.fields.length+1,
```

Reply:
Done :)